File: cs/Day13.cs
using System;
using System.Collections.Generic;
using System.IO;
using System.Linq;

namespace cs
{
    class Day13
    {
        public static void Part1()
        {
            List<List<bool>> paper = new List<List<bool>>();
            List<string> lines = File.ReadLines(@"../../../../../data/day13.txt").ToList().Select(line => line.Trim()).ToList();
            int folding_section_start = 0;

            for (int line_id = 0; line_id < lines.Count; line_id++)
            {
                folding_section_start++;
                string line = lines[line_id];
                if (line == "") break;

                List<int> ln = line.Split(",").Select(q => int.Parse(q)).ToList();
                int x = ln[0], y = ln[1];

                while (paper.Count <= y) paper.Add(new List<bool>());
                while (paper[y].Count <= x) paper[y].Add(false);

                paper[y][x] = true;
            }

            int longest = 0;
            foreach (List<bool> line in paper) longest = Math.Max(longest, line.Count);

            foreach (List<bool> line in paper) while (line.Count < longest) line.Add(false);

            string fold_line = lines[folding_section_start];
            string insterest = fold_line[(fold_line.LastIndexOf(" ") + 1)..];
            string fold_axis = insterest.Split("=")[0];
            int fold_index = int.Parse(insterest.Split("=")[1]);

            if (fold_axis == "x")
                for (int line = 0; line < paper.Count; line++)
                {
                    for (int index = 0; index < longest - fold_index - 1; index++)
                        paper[line][index] |= paper[line][longest - index - 1];
                    paper[line].RemoveRange(fold_index, fold_index + 1);
                }

            else if (fold_axis == "y")
            {
                for (int line = 0; line < paper.Count - fold_index - 1; line++)
                    for (int index = 0; index < longest; index++)
                        paper[line][index] |= paper[paper.Count - line - 1][index];
                paper.RemoveRange(fold_index, fold_index + 1);
            }

            int result = 0;
            foreach (List<bool> line in paper)
                foreach (bool c in line)
                    if (c) result++;

            Console.WriteLine("Day: 13 | Part: 1 | Result: " + result);
        }

        public static void Part2()
        {
            List<List<bool>> paper = new List<List<bool>>();
            List<string> lines = File.ReadLines(@"../../../../../data/day13.txt").ToList().Select(line => line.Trim()).ToList();
            int folding_section_start = 0;

            for (int line_id = 0; line_id < lines.Count; line_id++)
            {
                folding_section_start++;
                string line = lines[line_id];
                if (line == "") break;

                List<int> ln = line.Split(",").Select(q => int.Parse(q)).ToList();
                int x = ln[0], y = ln[1];

                while (paper.Count <= y) paper.Add(new List<bool>());
                while (paper[y].Count <= x) paper[y].Add(false);

                paper[y][x] = true;
            }

            int longest = 0;
            foreach (List<bool> line in paper) longest = Math.Max(longest, line.Count);

            foreach (List<bool> line in paper) while (line.Count < longest) line.Add(false);

            foreach (string fold_line in lines.Where(line => lines.IndexOf(line) >= folding_section_start))
            {
                longest = 0;
                foreach (List<bool> line in paper) longest = Math.Max(longest, line.Count);

                string insterest = fold_line[(fold_line.LastIndexOf(" ") + 1)..];
                string fold_axis = insterest.Split("=")[0];
                int fold_index = int.Parse(insterest.Split("=")[1]);

                if (fold_axis == "x")
                    for (int line = 0; line < paper.Count; line++)
                    {
                        for (int index = 0; index < longest - fold_index - 1; index++)
                            paper[line][index] |= paper[line][longest - index - 1];
                        paper[line].RemoveRange(fold_index, fold_index + 1);
                    }

                else if (fold_axis == "y")
                {
                    for (int line = 0; line < paper.Count - fold_index - 1; line++)
                        for (int index = 0; index < longest; index++)
                            paper[line][index] |= paper[paper.Count - line - 1][index];
                    paper.RemoveRange(fold_index, fold_index + 1);
                }
            }

            Console.WriteLine("Day: 13 | Part: 2 | Result:");
            foreach (List<bool> line in paper)
            {
                foreach (bool c in line)
                    Console.Write(c ? "#" : " ");
                Console.WriteLine();
            }
        }
    }
}

File: python/Day13.py
def part_1() -> None:
    paper: list[list[bool]] = []

    with open("../data/day13.txt", "r") as dFile:
        while (line := dFile.readline().strip()) != "":
            x, y = line.split(",")
            x, y = int(x), int(y)

            while y not in range(len(paper)):
                paper.append([])

            while x not in range(len(paper[y])):
                paper[y].append(False)

            paper[y][x] = True

        longest: int = max(len(line) for line in paper)
        for line in paper:
            while len(line) < longest:
                line.append(False)

        fold_line = dFile.readline()
        interest = fold_line[fold_line.strip().rfind(" ") + 1 :]
        fold_axis, fold_index = interest.split("=")
        fold_index = int(fold_index)

        if fold_axis == "x":
            for line in range(len(paper)):
                for index in range(longest - fold_index - 1):
                    paper[line][index] |= paper[line][longest - index - 1]
                paper[line] = paper[line][:fold_index]

        elif fold_axis == "y":
            for line in range(len(paper) - fold_index - 1):
                for index in range(longest):
                    paper[line][index] |= paper[len(paper) - line - 1][index]
            paper = paper[:fold_index]

        result = sum(1 for line in paper for char in line if char)

        print("Day: 13 | Part: 1 | Result:", result)


def part_2() -> None:
    paper: list[list[bool]] = []

    with open("../data/day13.txt", "r") as dFile:
        while (line := dFile.readline().strip()) != "":
            x, y = line.split(",")
            x, y = int(x), int(y)

            while y not in range(len(paper)):
                paper.append([])

            while x not in range(len(paper[y])):
                paper[y].append(False)

            paper[y][x] = True

        longest: int = max(len(line) for line in paper)
        for line in paper:
            while len(line) < longest:
                line.append(False)

        for fold_line in dFile.readlines():

            interest = fold_line[fold_line.strip().rfind(" ") + 1 :]
            fold_axis, fold_index = interest.split("=")
            fold_index = int(fold_index)

            longest = max(len(line) for line in paper)

            if fold_axis == "x":
                for line in range(len(paper)):
                    for index in range(longest - fold_index - 1):
                        paper[line][index] |= paper[line][longest - index - 1]
                    paper[line] = paper[line][:fold_index]

            elif fold_axis == "y":
                for line in range(len(paper) - fold_index - 1):
                    for index in range(longest):
                        paper[line][index] |= paper[len(paper) - line - 1][index]
                paper = paper[:fold_index]

        print("Day: 13 | Part: 2 | Result:")
        for line in paper:
            for char in line:
                print("#" if char else " ", end="")
            print()


if __name__ == "__main__":
    part_1()
    part_2()
